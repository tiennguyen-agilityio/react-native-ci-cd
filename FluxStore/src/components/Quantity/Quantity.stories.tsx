import {Alert} from 'react-native';
import {Meta, StoryObj} from '@storybook/react';

import Quantity from '.';

const meta: Meta<typeof Quantity> = {
  title: 'Components/Quantity',
  component: Quantity,
};

export default meta;

type Story = StoryObj<typeof Quantity>;

export const Default: Story = {
  args: {
    onChangeValue: (value: number) => Alert.alert('Quantity', `Current value: ${value}`),
  },
};

export const Max: Story = {
  args: {
    max: 5,
    onChangeValue: () => {},
  },
};
