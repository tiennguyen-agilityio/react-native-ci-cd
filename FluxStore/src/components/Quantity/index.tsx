import {memo, useCallback, useState} from 'react';
import {StyleSheet, View} from 'react-native';

import {useThemeStore} from '@/stores';
import {fontSizes} from '@/themes';

import Text from '../Text';
import {MinusIcon, AddIcon} from '../Icons';

interface QuantityProps {
  defaultValue?: number;
  max?: number;
  onChangeValue: (value: number) => void;
}

const Quantity = ({defaultValue = 0, max, onChangeValue}: QuantityProps) => {
  const {
    theme: {border},
  } = useThemeStore();

  const [number, setNumber] = useState<number>(defaultValue);

  const handleIncrement = useCallback(() => {
    const result = number + 1;

    setNumber(result);
    onChangeValue(result);
  }, [number, onChangeValue]);

  const handleDecrement = useCallback(() => {
    setNumber(number - 1);
    onChangeValue(number - 1);
  }, [number, onChangeValue]);

  return (
    <View style={[styles.wrapper, {borderColor: border.senary}]}>
      <MinusIcon disabled={number <= 1} onPress={handleDecrement} style={styles.icon} />
      <Text style={[styles.text, {color: border.senary}]} fontSize={fontSizes.tiny}>
        {number}
      </Text>
      <AddIcon
        style={styles.icon}
        disabled={Boolean(max && number === max)}
        onPress={handleIncrement}
      />
    </View>
  );
};

export default memo(Quantity);

const styles = StyleSheet.create({
  wrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 20,
    width: 63,
    height: 22,
    paddingHorizontal: 8,
  },
  icon: {
    height: 13,
    width: 13,
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    flex: 0,
  },
});
