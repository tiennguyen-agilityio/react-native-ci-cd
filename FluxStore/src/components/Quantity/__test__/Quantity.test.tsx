import {fireEvent, render, screen, waitFor} from '@testing-library/react-native';

import Quantity from '..';

describe('Quantity', () => {
  const onChangeValue = jest.fn();
  const props = {
    onChangeValue,
  };

  it('should render quantity component', () => {
    const {toJSON} = render(<Quantity {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onChangeValue when increment icon value clicked', async () => {
    render(<Quantity {...props} />);

    const incrementIcon = screen.getByTestId('add-icon');
    fireEvent.press(incrementIcon);

    await waitFor(() => {
      expect(onChangeValue).toHaveBeenCalled();
    });
  });

  it('calls onChangeValue when decrement icon value clicked', async () => {
    render(<Quantity {...props} defaultValue={3} />);

    const decrementIcon = screen.getByTestId('minus-icon');
    fireEvent.press(decrementIcon);

    await waitFor(() => {
      expect(onChangeValue).toHaveBeenCalled();
    });
  });

  it('does not decrement below 1', async () => {
    render(<Quantity {...props} defaultValue={1} />);

    const decrementIcon = screen.getByTestId('minus-icon');
    fireEvent.press(decrementIcon);

    await waitFor(() => {
      expect(onChangeValue).toHaveBeenCalled();
    });
  });

  it('not calls onChangeValue when increment icon value clicked', async () => {
    render(<Quantity {...props} defaultValue={3} max={3} />);

    const incrementIcon = screen.getByTestId('add-icon');
    fireEvent.press(incrementIcon);

    await waitFor(() => {
      expect(onChangeValue).toHaveBeenCalled();
    });
  });
});
