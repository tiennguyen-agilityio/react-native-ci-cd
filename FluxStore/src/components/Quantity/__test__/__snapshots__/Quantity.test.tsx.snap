// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity should render quantity component 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 20,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 22,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": 63,
      },
      {
        "borderColor": "#00000080",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "height": 13,
        "justifyContent": "center",
        "opacity": 1,
        "width": 13,
      }
    }
    testID="minus-icon"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={3}
      bbWidth={6}
      fill="none"
      focusable={false}
      height={3}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 3,
            "width": 6,
          },
        ]
      }
      vbHeight={3}
      vbWidth={6}
      width={6}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M0.100098 1.31445L5.89475 1.31445"
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          propList={
            [
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={
            {
              "payload": 2147483648,
              "type": 0,
            }
          }
          strokeWidth={1.5}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {
          "fontSize": 12,
        },
        [
          {
            "flex": 0,
          },
          {
            "color": "#00000080",
          },
        ],
      ]
    }
  >
    0
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "height": 13,
        "justifyContent": "center",
        "opacity": 1,
        "width": 13,
      }
    }
    testID="add-icon"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={8}
      bbWidth={8}
      fill="none"
      focusable={false}
      height={8}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 8,
            "width": 8,
          },
        ]
      }
      vbHeight={14}
      vbWidth={14}
      width={8}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M13 8H8v5c0 .55-.45 1-1 1s-1-.45-1-1V8H1c-.55 0-1-.45-1-1s.45-1 1-1h5V1c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
          fill={
            {
              "payload": 2147483648,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
